alkos/kernel/arch/x86_64/utils/serial/serial64.nasm

```diff
@@ -0,0 +1,163 @@
+          out     dx, al
+
+          ; DLM (COM1+1 = 0x3F9)
+          mov     dx, COM1 + 1
+          mov     al, 0x00
+          out     dx, al
+
+          ; Clear DLAB and set data bits (8 bits, no parity, one stop bit)
+          ; LCR (COM1+3 = 0x3FB)
+          mov     dx, COM1 + 3
+          mov     al, 0x03
+          out     dx, al
+
+          ; Enable FIFO, clear them, 14-byte threshold: FCR (COM1+2 = 0x3FA)
+          mov     dx, COM1 + 2
+          mov     al, 0xC7
+          out     dx, al
+
+          ; Enable IRQs, RTS/DSR: MCR (COM1+4 = 0x3FC)
+          mov     dx, COM1 + 4
+          mov     al, 0x0B
+          out     dx, al
+
+          ; Restore RDX and RAX registers
+          pop     rdx
+          pop     rax
+          ret
+
+; -----------------------------------------------------------------------------
+; Function: serial_putchar
+; Description:
+;     Sends a single character to the serial port (COM1).
+; Parameters:
+;     char c  - The character to send. Passed in the lower 8 bits of RDI.
+; Returns:
+;     void
+; -----------------------------------------------------------------------------
+serial_putchar:
+          push    rbp
+          mov     rbp, rsp
+          push    rax
+          push    rbx
+          push    rdx
+
+          ; Move the character from RDI (first argument) to AL
+          mov     al, dil
+
+          ; Wait until the transmit buffer is empty
+.wait_ready:
+          ; Read Line Status Register (LSR) at COM1+5 = 0x3FD
+          mov     dx, COM1 + 5
+          in      al, dx
+          test    al, EMPTY
+          jz      .wait_ready
+
+          ; Send the character to COM1
+          mov     al, dil
+          mov     dx, COM1
+          out     dx, al
+
+          pop     rdx
+          pop     rbx
+          pop     rax
+          pop     rbp
+          ret
+
+; -----------------------------------------------------------------------------
+; Function: serial_puts
+; Description:
+;     Sends a null-terminated string to the serial port (COM1).
+; Parameters:
+;     const char* str  - Pointer to the null-terminated string. Passed as
+;                        the first argument in RDI.
+; Returns:
+;     void
+; -----------------------------------------------------------------------------
+serial_puts:
+          push    rbp
+          mov     rbp, rsp
+          push    rbx
+          push    rax
+
+          ; Move the string pointer from RDI (first argument) to RBX
+          mov     rbx, rdi
+
+.loop:
+          ; Load the current character into AL
+          mov     al, byte [rbx]
+          cmp     al, NULL
+          je      .done
+
+          ; Move the character to DIL (lower 8 bits of RDI) for serial_putchar
+          mov     dil, al
+          call    serial_putchar
+
+          ; Move to the next character in the string
+          inc     rbx
+          jmp     .loop
+
+.done:
+          pop     rax
+          pop     rbx
+          pop     rbp
```

Comment:
do we really need to pop push all those registers? As we want to follow some call convetions, some registers can be changed without pushing and secondly both functions (putchar and puts) could cooperate with each other in this matter

Reply:
YES.

I may change it in the future but for now God no. Don't let this edit the registers.

Reply:
in future lets follow some simple conventions because integrating other code od modyfing it is literally impossible without that -> sysV convention